commit ea09eda249fbf39024a2d1b18b0e99183fc5bf2d
Author: finagolfin
Date:   Wed Apr 22 01:08:28 2026

    Revert the new `#available(Android <API>, *)` instead to look for `backtrace()`

    This is needed only because we're missing the right nullability annotations
    on the Termux execinfo.h and can be thrown out once that's corrected.
    
diff --git a/swift-testing/Sources/Testing/SourceAttribution/Backtrace.swift b/swift-testing/Sources/Testing/SourceAttribution/Backtrace.swift
index 3c6e412b..7cf7158e 100644
--- a/swift-testing/Sources/Testing/SourceAttribution/Backtrace.swift
+++ b/swift-testing/Sources/Testing/SourceAttribution/Backtrace.swift
@@ -44,6 +44,16 @@ public struct Backtrace: Sendable {
     self.addresses = addresses.map { Address(UInt(bitPattern: $0)) }
   }
 
+#if os(Android) && !SWT_NO_DYNAMIC_LINKING
+  /// The `backtrace()` function.
+  ///
+  /// This function was added to Android with API level 33, which is higher than
+  /// our minimum deployment target, so we look it up dynamically at runtime.
+  private static let _backtrace = symbol(named: "backtrace").map {
+    castCFunction(at: $0, to: (@convention(c) (UnsafeMutablePointer<UnsafeMutableRawPointer?>, CInt) -> CInt).self)
+  }
+#endif
+
   /// Get the current backtrace.
   ///
   /// - Parameters:
@@ -66,11 +76,11 @@ public struct Backtrace: Sendable {
 #if SWT_TARGET_OS_APPLE
       initializedCount = backtrace_async(addresses.baseAddress!, addresses.count, nil)
 #elseif os(Android)
-      if #available(Android 33, *) {
-        initializedCount = addresses.withMemoryRebound(to: UnsafeMutableRawPointer.self) { addresses in
-          .init(clamping: backtrace(addresses.baseAddress!, .init(clamping: addresses.count)))
-        }
+#if !SWT_NO_DYNAMIC_LINKING
+      if let _backtrace {
+        initializedCount = .init(clamping: _backtrace(addresses.baseAddress!, .init(clamping: addresses.count)))
       }
+#endif
 #elseif os(Linux) || os(FreeBSD) || os(OpenBSD)
       initializedCount = .init(clamping: backtrace(addresses.baseAddress!, .init(clamping: addresses.count)))
 #elseif os(Windows)
